scripts/integrity_check.py

```diff
@@ -0,0 +1,141 @@
+#
+# Purpose
+#
+# This script requires python 3. This script checks for undesired changes
+# to the source code, including file permission changes, presence of tabs,
+# non-Unix line endings, trailing whitespace, and TODO comments.
+# Note: must be run from Mbed TLS root.
+
+import os
+import argparse
+import logging
+
+
+class IntegrityChecker(object):
+
+    def __init__(self, log_file):
+        self.check_repo_path()
+        self.logger = None
+        self.setup_logger(log_file)
+        self.source_files_to_check = (
+            ".c", ".h", ".function", ".data",
+            ".md", "Makefile", "CMakeLists.txt"
+        )
+        self.script_file_types = (".sh", ".pl", ".py")
+        self.permission_issues = []
+        self.end_of_file_newline_issues = []
+        self.line_ending_issues = {
+            "report_heading": "Non Unix line endings:", "files": {}
+        }
+        self.trailing_whitespace = {
+            "report_heading": "Trailing whitespace:", "files": {}
+        }
+        self.tab_issues = {
+            "report_heading": "Tabs present:", "files": {}
+        }
+        self.todo_issues = {
+            "report_heading": "TODO present:", "files": {}
+        }
+
+    def check_repo_path(self):
+        if not __file__ == os.path.join(".", "scripts", "integrity_check.py"):
+            raise Exception("Must be run from Mbed TLS root")
+
+    def setup_logger(self, log_file, level=logging.INFO):
+        self.logger = logging.getLogger()
+        self.logger.setLevel(level)
+        if log_file:
+            handler = logging.FileHandler(log_file)
+            self.logger.addHandler(handler)
+        else:
+            console = logging.StreamHandler()
+            self.logger.addHandler(console)
+
+    def check_file_permissions(self, filepath):
+        if not (os.access(filepath, os.X_OK) ==
+                filepath.endswith(self.script_file_types)):
+            self.permission_issues.append(filepath)
+
+    def check_file_content(self, filepath):
+        with open(filepath, "r") as f:
+            self.line_ending_issues["files"][filepath] = []
+            self.trailing_whitespace["files"][filepath] = []
+            self.tab_issues["files"][filepath] = []
+            self.todo_issues["files"][filepath] = []
+            allow_trailing_whitespace = filepath.endswith(".md")
+            allow_tabs = filepath.endswith("Makefile")
+            line = None
+            for i, line in enumerate(iter(f.readline, "")):
+                if "\r" in line:
+                    self.line_ending_issues["files"][filepath].append(i + 1)
+                if (not allow_trailing_whitespace and
+                        line.rstrip("\r\n") != line.rstrip()):
+                    self.trailing_whitespace["files"][filepath].append(i + 1)
+                if not allow_tabs and "\t" in line:
+                    self.tab_issues["files"][filepath].append(i + 1)
+                if "TODO" in line:
```

Comment:
How does this script cope with itself? It contains `TODO`.

Reply:
The script originally wasn't checking `.py` files, so it didn't check itself. I've added an exemption in the rewrite.